modules/calib3d/src/ippe.cpp

```diff
@@ -0,0 +1,1122 @@
+
+    int n = _objectPoints.rows() * _objectPoints.cols();
+
+    _canonicalObjPoints.create(1, n, CV_64FC2);
+    _MmodelPoints2Canonical.create(4, 4, CV_64FC1);
+
+    Mat objectPoints = _objectPoints.getMat();
+    Mat canonicalObjPoints = _canonicalObjPoints.getMat();
+
+    Mat UZero(3, n, CV_64FC1);
+
+    double xBar = 0;
+    double yBar = 0;
+    double zBar = 0;
+    bool isOnZPlane = true;
+    for (int i = 0; i < n; i++)
+    {
+        double x, y, z;
+        if (objType == CV_32FC3)
+        {
+            x = static_cast<double>(objectPoints.at<Vec3f>(i)[0]);
+            y = static_cast<double>(objectPoints.at<Vec3f>(i)[1]);
+            z = static_cast<double>(objectPoints.at<Vec3f>(i)[2]);
+        }
+        else
+        {
+            x = objectPoints.at<Vec3d>(i)[0];
+            y = objectPoints.at<Vec3d>(i)[1];
+            z = objectPoints.at<Vec3d>(i)[2];
+        }
+
+        if (abs(z) > IPPE_SMALL)
+        {
+            isOnZPlane = false;
+        }
+
+        xBar += x;
+        yBar += y;
+        zBar += z;
+
+        UZero.at<double>(0, i) = x;
+        UZero.at<double>(1, i) = y;
+        UZero.at<double>(2, i) = z;
+    }
+    xBar = xBar / static_cast<double>(n);
+    yBar = yBar / static_cast<double>(n);
+    zBar = zBar / static_cast<double>(n);
+
+    for (int i = 0; i < n; i++)
+    {
+        UZero.at<double>(0, i) -= xBar;
+        UZero.at<double>(1, i) -= yBar;
+        UZero.at<double>(2, i) -= zBar;
+    }
+
+    Mat MCenter(4, 4, CV_64FC1);
+    MCenter.setTo(0);
+    MCenter.at<double>(0, 0) = 1;
+    MCenter.at<double>(1, 1) = 1;
+    MCenter.at<double>(2, 2) = 1;
+    MCenter.at<double>(3, 3) = 1;
+
+    MCenter.at<double>(0, 3) = -xBar;
+    MCenter.at<double>(1, 3) = -yBar;
+    MCenter.at<double>(2, 3) = -zBar;
+
+    if (isOnZPlane)
+    {
+        //MmodelPoints2Canonical is given by MCenter
+        MCenter.copyTo(_MmodelPoints2Canonical);
+        for (int i = 0; i < n; i++)
+        {
+            canonicalObjPoints.at<Vec2d>(i)[0] = UZero.at<double>(0, i);
+            canonicalObjPoints.at<Vec2d>(i)[1] = UZero.at<double>(1, i);
+        }
+    }
+    else
+    {
+        Mat UZeroAligned(3, n, CV_64FC1);
+        Mat R; //rotation that rotates objectPoints to the plane z=0
+
+        if (!computeObjextSpaceR3Pts(objectPoints,R))
+        {
+            //we could not compute R, problably because there is a duplicate point in {objectPoints(0),objectPoints(1),objectPoints(2)}.
+            //So we compute it with the SVD (which is slower):
+            computeObjextSpaceRSvD(UZero,R);
+        }
+
+        UZeroAligned = R * UZero;
+
+        for (int i = 0; i < n; i++)
+        {
+            canonicalObjPoints.at<Vec2d>(i)[0] = UZeroAligned.at<double>(0, i);
+            canonicalObjPoints.at<Vec2d>(i)[1] = UZeroAligned.at<double>(1, i);
+            //or
+            //TODO:
+//            CV_DbgAssert(abs(UZeroAligned.at<double>(2, i)) <= IPPE_SMALL);
```

Comment:
Should we keep the `CV_DbgAssert` or maybe find a way to return false to not interrupt the main program?

Reply:
I believe, `CV_Error()` with `Error::StsNoConv` is fine here.

> not interrupt the main program

"main program" should expect that OpenCV C++ functions can raise exceptions and handle that cases.